File: Part I: Introduction to Probability/the-logic-of-uncertainty.R
# What is the probability of rolling 20 three times in a row on a 20-sided die?
sides <- 20
times <- 3
event.prob <- (1/sides)^times

# Weather report says there's a 10% chance of rain and you forget your umbrella 50%.
# What is the chance you get caught in the rain without your umbrella?
rain.prob <- 0.1
umb.prob <- 0.5
rain.no.umb <- rain.prob*umb.prob

# Raw eggs have a 1/20000 chance of having salmonella.
# If you eat two raw eggs, what is the probability you ate an egg with salmonella?
times <- 2
prob <- 1/20000
salm.prob <- (prob*times)-prob^times

# What is the probability of getting two heads in two coin tosses or
# rolling three sixes in three 6-sided dice rolls?
coin.prob <- 1/2
dice.prob <- 1/6
rolls <- 3
flips <- 2
heads.sixes <- coin.prob^flips+dice.prob^rolls-coin.prob^flips*dice.prob^rolls

File: Part I: Introduction to Probability/thelogicofuncertainty.py
# What is the probability of rolling 20 three times in a row on a 20-sided die?
def event_prob(prob_one_event, event_occurs):
    return prob_one_event ** event_occurs


# Weather report says there's a 10% chance of rain and you forget your umbrella 50%.
# What is the chance you get caught in the rain without your umbrella?
def and_prob(*events):
    prob = 1
    for event in events:
        prob *= event
    return prob


# Raw eggs have a 1/20000 chance of having salmonella.
# If you eat two raw eggs, what is the probability you ate an egg with salmonella?
def or_prob(times, *events):
    prob: int = 0
    mutual: int = 1
    for event in events:
        prob += event * times
        mutual *= event_prob(event, times)
    return prob - mutual


# What is the probability of getting two heads in two coin tosses or
# rolling three sixes in three 6-sided dice rolls?
def or_two_prob(times1, times2, event1, event2):
    return event_prob(event1, times1) + event_prob(event2, times2) - \
           and_prob(event_prob(event1, times1), event_prob(event2, times2))


if __name__ == "__main__":
    one = event_prob(0.05, 3)
    two = and_prob(0.1, 0.5)
    three = or_prob(2, (1/20000))
    four = or_two_prob(2, 3, 1/2, 1/6)
